proposals/0000-no-implicit-binds.rst

```diff
@@ -0,0 +1,446 @@
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     data D @x (y :: x) (z :: y) where -- OK, someday
+
+   But today we have to this::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     data D :: forall x. forall (y :: x) -> y -> Type where
+
+   Or in 8.10 alternatively this::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type D :: forall x. forall (y :: x) -> y -> Type
+     data D where -- OK
+
+#. ::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type family F (y :: x) (z :: y) where -- error: `x` is not bound, `y` and `z` are fine
+
+   Could *someday* be be rewritten as::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type family F @x (y :: x) (z :: y) where -- OK, someday
+
+   But today we have to this::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type family F :: forall x. forall (y :: x) -> y -> Type where
+
+   or in 8.10 alternatively this::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type F :: forall x. forall (y :: x) -> y -> Type
+     type family F where -- OK
+
+#. ::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     class Eq a => C (y :: x) (z :: y) where -- error: `x` is not bound, `y` and `z` are fine
+
+   Could *someday* be be rewritten as::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     class Eq a => C @x (y :: x) (z :: y) where -- OK, someday
+
+   But in 8.10 we have to this::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     type C :: forall x. forall (y :: x) -> y -> Constraint
+     class Eq a => C y z where -- OK
+
+   Note that since there is no ``class F :: ...`` syntax analogous to ``data F :: ...``,
+   so ``-XStandaloneKindSignatures`` are the only way to write explicitly kind-polymorphic classes.
+
+Note that the variables to the left of the ``::`` are are deemed explicit bindings analogous to ``f (y :: x) (z :: z) = ...`` and permitted.
+However ``x`` to the right of the ``::`` is a use, not otherwise bound, and thus implicit binding today.
+It is not permitted as-is, and must be explicitly bound or discarded as done in the working alternatives.
+
+Empty `forall` "desugar"
+~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
You should probably move this subsection to the `Cost` and drawback as it is a justification that this proposal oughtn't be costly to implement.

Reply:
Moved.